File: 01-basic-auth/go/internal/domain/repository.go
package domain

// UserRepository — контракт хранилища пользователей.
//
// UseCase зависит от этого интерфейса (а не от конкретной БД),
// поэтому мы можем легко заменить in-memory реализацию на Postgres/Redis.
type UserRepository interface {
	Create(user *User) error
	GetByID(id string) (*User, error)
	GetByEmail(email string) (*User, error)
	Delete(id string) error
}

File: 01-basic-auth/node/src/domain/repository.ts
import { User } from './user';

/**
 * UserRepository — контракт хранилища пользователей.
 *
 * UseCase зависит от интерфейса, а не от конкретной реализации,
 * поэтому можно заменить in-memory на Postgres/Redis без изменения бизнес-логики.
 */
export interface UserRepository {
  create(user: User): Promise<void>;
  getById(id: string): Promise<User | null>;
  getByEmail(email: string): Promise<User | null>;
  delete(id: string): Promise<void>;
}
